File: Cpp-version/Plant.cpp
#include "Plant.h"

Plant::Plant(int x, int y, World& world) 
	: Organism(x,y,world) {
	ini = 0;
}
Plant::Plant(int str, int ini, int xPos, int yPos, int age, int oId, bool moved, World& world)
	:Organism(str, ini, xPos, yPos, age, oId, moved, world) {}

void Plant::TakeAction() {
	if (!moved && age > 0)
	{
		int rNum, spreadsAttempted = 0;
		while (spreadsAttempted != spreadCount)
		{
			rNum = rand() % 100;
			cout << name << " from tile (" << xPos << ", " << yPos << ") tried spreading";
			if (rNum >= 99 - spreadChance)
				Spread();
			else
				cout << ", but failed" << endl;
			spreadsAttempted++;
		}
		age++;
	}
}

void Plant::Win(Organism& otherCreature) {
	cout << " and " << otherCreature.GetName() << " died while eating it! Both perished" << endl;
	world.RmOrganism(otherCreature.GetOId());
	world.RmOrganism(oId);
	otherCreature.SetAge(DEAD);
	age = DEAD;
}

bool Plant::Defended(Organism& attackingCreature) {
	if (str > attackingCreature.GetS())
		return true;
	return false;
}

void Plant::Spread() {
	bool triedLeft = false, triedRight = false, triedUp = false, triedDown = false;
	while (true)
	{
		int changeX = 0, changeY = 0, rNum;
		rNum = (rand() % 100) % 4;
		if (rNum == 0) {
			if (triedLeft)
				continue;
			else
				triedLeft = true;
			changeX = -1;
		}
		else if (rNum == 1) {
			if (triedRight)
				continue;
			else
				triedRight = true;
			changeX = 1;
		}
		else if (rNum == 2) {
			if (triedUp)
				continue;
			else
				triedUp = true;
			changeY = -1;
		}
		else {
			if (triedDown)
				continue;
			else
				triedDown = true;
			changeY = 1;
		}
		if (STAYS_IN_BOUNDS)
		{
			Organism* potentialCollision = world.LookForCollision(xPos + changeX, yPos + changeY);
			if (potentialCollision == nullptr)
			{
				Plant* newPlant = MakeNewP(xPos + changeX, yPos + changeY);
				world.AddOrganism(newPlant);
				cout << " and spread to tile (" << xPos + changeX << ", " << yPos + changeY << ")" << endl;
				break;
			}
		}
		else {
			changeX = 0;
			changeY = 0;
		}
		if (triedDown && triedUp && triedLeft && triedRight)
		{
			cout << ", but had no adjacent tiles to spread to" << endl;
			break;
		}
	}
}
void Plant::Print() {
	world.SetMap(xPos, yPos, sign);
}

Plant::~Plant() {}

File: Python-version/plant.py
from organism import Organism
from abc import ABC, abstractmethod
import random

class Plant(Organism):
    def __init__(self, x, y, world, oid=None, s=None, i=None, a=None, m=None):
        super().__init__(x, y, world, oid, str, i, a, m)
        self.color="black"

        self.spread_count=1
        self.spread_chance=0

    @abstractmethod
    def MakeNewP(self,x,y):
        pass

    def print(self):
        self.world.SetMap(self.xPos, self.yPos, self.sign, self.color)

    def Win(self,othercreature):
        print(f" and {othercreature.GetName()} died while eating it! Both perished\n")
        self.log = f"{self.log} and {othercreature.GetName()} died while eating it! Both perished\n"
        self.world.RmOrganism(self.GetOId())
        self.world.RmOrganism(othercreature.GetOId())

        self.world.AddLog(othercreature.GetLog())
        self.world.AddLog(self.log)
        othercreature.ResetLog()
        othercreature.SetAge(-1)
        log=" "
        self.age=-1

    def Defended(self,attackingcreature):
        if (self.str > attackingcreature.GetS()):
            return True
        return False

    def TakeAction(self):
        if self.age!=-1 and not self.moved:
            spreads_attempted=0
            while(spreads_attempted!=self.spread_count):#tu moze byc blad bo spreadcount jest tu ustawiony na 0
                rand=random.Random()
                rnum=rand.randint(0,99)
                print(f"{self.name} from tile ({self.xPos}, {self.yPos}) tried spreading ",end="")
                self.log = f"{self.log} {self.name} from tile ({self.xPos}, {self.yPos}) tried spreading"
                if rnum>=(99-self.spread_chance):
                    self.Spread()
                else:
                    print(f", but failed\n")
                    self.log=f"{self.log}, but failed\n"
                spreads_attempted+=1
            self.age+=1

    def Spread(self):
        triedLeft = False
        triedRight = False
        triedUp = False
        triedDown = False
        while (True):
            rand = random.Random()
            changeX = 0
            changeY = 0
            rNum = rand.randint(0, 3)
            if rNum == 0:
                if triedLeft:
                    continue
                else:
                    triedLeft = True
                changeX = -1
            elif rNum == 1:
                if triedRight:
                    continue
                else:
                    triedRight = True
                changeX = 1
            elif rNum == 2:
                if triedUp:
                    continue
                else:
                    triedUp = True
                changeY = -1
            else:
                if triedDown:
                    continue
                else:
                    triedDown = True
                changeY = 1
            if self.xPos + changeX < self.world.GetW() and self.yPos + changeY < self.world.GetH() and self.xPos + changeX >= 0 and self.yPos + changeY >= 0:
                potential_collision = self.world.LookForCollision(self.xPos + changeX, self.yPos + changeY)
                if potential_collision is None:
                    new_plant=self.MakeNewP(self.xPos+changeX,self.yPos+changeY)
                    self.world.AddOrganism(new_plant)
                    tmpx = self.xPos + changeX
                    tmpy = self.yPos + changeY
                    print(f" and spread to tile({tmpx}, {tmpy})\n")
                    break
            if triedDown and triedRight and triedUp and triedLeft:
                print(f", but had no adjacent tiles to spread to\n")
                break


